### PR TITLE
Use a QuadTree to optimise rendering

### PR DESCRIPTION
**Pull request by AgmasGold**
This PR attempts to implement the use of a QuadTree within AnnoCanvas.

By using a [Quad Tree](https://en.wikipedia.org/wiki/Quadtree) structure, we can speed up collision detection, as we can easily throwaway parts of the canvas that will never intersect. It means we can do collision detection without needing to check every single object in the PlacedObjects list.

Additionally, a "viewport" can be added, which would mean we would only have to process what was on screen at the time, and functions like drag all or zooming can be more easily implemented - rather than moving all the objects themselves, we can just shift the viewport, by drawing all objects with a specific offset. 

Alongside the rendering improvements, this change should also allow us to add scroll bars to the canvas when its wrapped in a `ScrollViewer`.

To Do:
* [x] Implement collision checking
* [x] Check if _oldObject position bug is fixed
* [x] Fix issue with viewport rect
* [x] Cache `GridRect` in `LayoutObject` for use in `GetObjectAt`
* [x] Handle items which are offscreen but still project an influence radius/range
* [ ] Write tests for QuadTree
* [x] ~Calling "Normalise" on large layouts is now very slow as we need to reindex the entire QuadTree.~ Solved by adding an overload to the QuadTree that allows deriving bounds from the objects themselves during a re-index operation.
* [x] Implement ScrollViewer support
  * [x] Align grid to viewport
  * [x] Invert scroll direction
  * [x] Invert panning direction
  * [x] Update scroll bars on resize
* [x] Handle when placing an object outside the quadtree (increase bounds and re-index)

### CODE REVIEW

**Comment by AgmasGold** [comment]
This change is now ready. The final thing I added was a couple of options to the General Settings page to invert the panning and scrolling direction.
image

File: AnnoDesigner.Core/DataStructures/QuadTree.cs [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+        /// </summary>
+        /// <param name="bounds">The bounds to check</param>
+        /// <returns></returns>
+        public IEnumerable<T> GetItemsIntersecting(Rect bounds)
+        {
+            var items = new List<T>(previousCount);
+            root.GetItemsIntersecting(items, bounds);
+            //using Count() here is fine as the underlying collection is a List, which implements ICollection
+            //https://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,1314
+            previousCount = items.Count();
```
Comment (FroggieFrog) [inline AnnoDesigner.Core/DataStructures/QuadTree.cs:375]: Why not just using `previousCount = items.Count;`?

File: AnnoDesigner/ViewModels/MainViewModel.cs [hunk 2]
```diff
@@ -1086,7 +1086,7 @@ private void ExecuteExportImageSub(bool exportZoom, bool exportSelection)
         /// <param name="exportSelection">indicates whether selection and influence highlights should be rendered</param>
         private void RenderToFile(string filename, int border, bool exportZoom, bool exportSelection, bool renderStatistics)
         {
-            if (AnnoCanvas.PlacedObjects.Count == 0)
+            if (AnnoCanvas.PlacedObjects.Count() == 0)
```
[the remark this answers is not in the document]
Comment (AgmasGold) [reply AnnoDesigner/ViewModels/MainViewModel.cs:1089]: `.Count()` should be a little faster than `Any()`. It doesn't usually matter, but as this is used in key performance code, the small optimisations make a difference.

File: AnnoDesigner.Core/DataStructures/QuadTree.cs [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+
+namespace AnnoDesigner.Core.DataStructures
+{
+    /// <summary>
+    /// Creates a new <see cref="QuadTree{T}"/>.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class QuadTree<T> : IEnumerable<T>
+    {
+        public class Quadrant
```
[the remark this answers is not in the document]
Comment (AgmasGold) [reply AnnoDesigner.Core/DataStructures/QuadTree.cs:19]: Yep, should be private.

File: AnnoDesigner.Core/DataStructures/QuadTree.cs [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+        {
+            /// <summary>
+            /// Indicates if current cached data (<see cref="count"/> and <see cref="itemCache"/>) is up to date.
+            /// Returns <see langword="true"/> when cached data is up to date.
+            /// </summary>
+            bool isDirty;
```
[the remark this answers is not in the document]
Comment (AgmasGold) [reply AnnoDesigner.Core/DataStructures/QuadTree.cs:25]: Done.

File: AnnoDesigner.Core/DataStructures/QuadTree.cs [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+            /// </summary>
+            IEnumerable<(T Item, Rect Bounds)> itemCache;
+
+            /// <summary>
+            /// Holds a list of all the items in this quadrant.
+            /// </summary>
+            /// <remarks>
+            /// Stored as a list as items as any items that overlap multiple child quadrants will be stored here.
+            /// </remarks>
+            public List<(T Item, Rect Bounds)> Items { get; }
```
[the remark this answers is not in the document]
Comment (AgmasGold) [reply AnnoDesigner.Core/DataStructures/QuadTree.cs:58]: This holds just the list of items directly held within this quadrant (e.g items that do not fit completely within a sub-quadrant). `itemCache` holds everything in this quadrant and the sub-quadrants "under" it.

File: AnnoDesigner.Core/Layout/Helper/StatisticsCalculationHelper.cs [hunk 6]
```diff
@@ -10,16 +10,22 @@ namespace AnnoDesigner.Core.Layout.Helper
...
-            if (placedObjects == null)
+            if (objects == null)
             {
                 return null;
             }
 
             var result = new StatisticsCalculationResult();
 
-            if (!placedObjects.Any())
+            if (objects.Count() == 0)
```
[the remark this answers is not in the document]
Comment (AgmasGold) [reply AnnoDesigner.Core/Layout/Helper/StatisticsCalculationHelper.cs:28]: `.Count() == 0` is slightly quicker than `Any()` for `IEnumerable`s where the underlying collection implements `ICollection` (which is all the collections we work with). Not something to usually be bothered about, but every bit of performance counts for this code.

File: AnnoDesigner/AnnoCanvas.xaml.cs [hunk 7]
```diff
@@ -637,50 +727,105 @@ private void AppSettings_SettingsChanged(object sender, EventArgs e)
...
+            var objectsToDraw = PlacedObjects.GetItemsIntersecting(_viewport.Absolute).ToList();
+            //borderless objects should be drawn first.
+            var borderlessObjects = objectsToDraw.Where(_ => _.WrappedAnnoObject.Borderless).ToList();
+            var borderedObjects = objectsToDraw.Where(_ => !_.WrappedAnnoObject.Borderless).ToList();
 
             // draw placed objects            
-            RenderObjectList(drawingContext, PlacedObjects, useTransparency: false);
+            RenderObjectList(drawingContext, borderlessObjects, useTransparency: false);
+            RenderObjectList(drawingContext, borderedObjects, useTransparency: false);
+            //RenderObjectList(drawingContext, objectsToDraw, useTransparency: false);
```
[the remark this answers is not in the document]
Comment (AgmasGold) [reply AnnoDesigner/AnnoCanvas.xaml.cs:828]: Done.

File: AnnoDesigner/AnnoCanvas.xaml.cs [hunk 8]
```diff
@@ -1232,6 +1554,72 @@ private void LoadObjectBorderLineColor()
...
+        /// <param name="oldPositions"></param>
+        /// <param name="newPositions"></param>
+        private void UpdateObjectPositions(IEnumerable<(LayoutObject, Rect)> oldPositions, IEnumerable<(LayoutObject, Rect)> newPositions)
+        {
+            //The IEnumerable sequences passed in as parameters sometimes get GC'd between calls when using MouseMode.DragAll, 
+            //as the objects are not referenced anywhere between the end of the foreach loop and the AddRange call (the variables
+            //themselves do not count as references due to IEnumerable lazy evaluation).
+            //By materializing the lists early, we ensure that references are maintained.
+            var materialisedOld = oldPositions.ToList();
+            var materialisedNew = newPositions.ToList();
```
[the remark this answers is not in the document]
Comment (AgmasGold) [reply AnnoDesigner/AnnoCanvas.xaml.cs:1572]: Good point - I've updated the method signature.

**Comment by AgmasGold** [comment]
Can we close #43 if this PR gets merged? It would still be awkward, but scrollbars would allow a touchpad to move around the canvas.

File: AnnoDesigner.Core/DataStructures/QuadTree.cs [hunk 9]
```diff
@@ -42,18 +42,19 @@ public class Quadrant
...
-            IEnumerable<(T Item, Rect Bounds)> itemCache;
+            private IEnumerable<(T Item, Rect Bounds)> itemCache;
 
             /// <summary>
             /// Holds a list of all the items in this quadrant.
             /// </summary>
             /// <remarks>
-            /// Stored as a list as items as any items that overlap multiple child quadrants will be stored here.
+            /// Stored as a list of items as any item that overlaps multiple child quadrants will be stored here.
+            /// This list differs to the itemCache, as this only contains items that do not fit completely into a child quadrant.
```
Comment (FroggieFrog) [inline AnnoDesigner.Core/DataStructures/QuadTree.cs:57]: So `Items` could be `ItemsInQuadrant` and `itemCache` could be `itemsInQuadrantIncludingChildren`?

**Review by FroggieFrog — APPROVED** [review]
I approve the structure of the code and the functionality.
I still have limited knowledge of a quadtree and viewboxes. But it seems to work 😁 

Good job 🎉

**Pull request merged**